File: device-app/ios-object-c/helloUms/helloUms/hello/ViewController.m
#import "ViewController.h"
#import <RationalOwl/MinervaManager.h>
#import <RationalOwl/Result.h>

#import "ViewController.h"

#import "UmsApi.h"
#import "PushAppProto.h"

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    MinervaManager *mgr = [MinervaManager getInstance];
    [mgr setDeviceRegisterResultDelegate:self];
    [mgr setMessageDelegate:self];
}


- (IBAction)btnReg:(id)sender {
    // airkorea
    NSString *gateHost = @"210.99.81.117";
    NSString *serviceId = @"SVC1829c8fd-920e-444d-805d-ef04e96c4811"; //@"696c
    //NSString *gateHost = @"linux.rationalowl.kr"; //@"ums.rationalowl.com"; //
    //NSString *serviceId = @"SVCf3173aae-d166-4199-b266-3aa78f601508"; //@"696c8a6aa7e44903b02e3948714d3887";
    MinervaManager *mgr = [MinervaManager getInstance];
    [mgr registerDevice:gateHost
              serviceId:serviceId
         deviceRegName:@"my Ios helloUms app"];
}

- (IBAction)btnUnreg:(id)sender {
    // airkorea
    NSString *serviceId = @"SVC1829c8fd-920e-444d-805d-ef04e96c4811";
    //NSString *serviceId = @"SVCf3173aae-d166-4199-b266-3aa78f601508";
    
    MinervaManager *mgr = [MinervaManager getInstance];
    [mgr unregisterDevice:serviceId];
}

#pragma mark - DeviceRegisterResultDelegate

- (void)onRegisterResult:(int32_t)resultCode
              resultMsg:(NSString *)resultMsg
            deviceRegId:(NSString *)deviceRegId {

    NSLog(@"onRegisterResult %d", resultCode);
    NSString *msg = [NSString stringWithFormat:@"%@ registration id: %@", resultMsg, deviceRegId];
    NSLog(@"%@", msg);

    if (resultCode == RESULT_OK || resultCode == RESULT_DEVICE_ALREADY_REGISTERED) {
        NSLog(@"rationalOwl register success!!!");
        //NSString *fDeviceRegId = deviceRegId;
        // airkorea
        [UmsApi callInstallUmsApp:@"9139c4e4-4fa3-47b1-bc6b-f4921fc54214"
                                   deviceRegId:deviceRegId
                                      phoneNum:nil
                                    appUserId:nil
                                           name:nil
                                    completion:^(NSData *data, NSError *error) {
            if (error) {
                NSLog(@"API 호출 실패: %@", error.localizedDescription);
                return;
            }

            NSString *responseString = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
            NSLog(@"응답 데이터: %@", responseString);

            NSError *jsonError = nil;
            NSDictionary *jsonDict = [NSJSONSerialization JSONObjectWithData:data options:0 error:&jsonError];
            if (jsonError) {
                NSLog(@"디코딩 오류: %@", jsonError.localizedDescription);
                return;
            }

            PushAppInstallRes *res = [[PushAppInstallRes alloc] initWithDictionary:jsonDict];
            if (res.rc == RESULT_OK) {
                self.mDeviceRegId = deviceRegId;
                NSLog(@"단말앱 등록 성공");
            } else {
                NSLog(@"에러 코드: %d, 사유: %@", res.rc, res.cmt);
            }
        }];
    } else {
        NSLog(@"단말앱 등록 에러: %@", resultMsg);
    }
}

- (void)onUnregisterResult:(int32_t)resultCode resultMsg:(NSString *)resultMsg {
    if (resultCode == RESULT_OK) {
        if (!self.mDeviceRegId) {
            NSLog(@"등록 아이디 없음");
            return;
        }
        // airkorea
        [UmsApi callUnregisterUmsApp:@"9139c4e4-4fa3-47b1-bc6b-f4921fc54214"
                                     deviceRegId:self.mDeviceRegId
                                      completion:^(NSData *data, NSError *error) {
            if (error) {
                NSLog(@"API 호출 실패: %@", error.localizedDescription);
                return;
            }

            NSError *jsonError = nil;
            NSDictionary *jsonDict = [NSJSONSerialization JSONObjectWithData:data options:0 error:&jsonError];
            if (jsonError) {
                NSLog(@"디코딩 오류: %@", jsonError.localizedDescription);
                return;
            }

            PushAppUnregUserRes *res = [[PushAppUnregUserRes alloc] initWithDictionary:jsonDict];
            if (res.rc == RESULT_OK) {
                self.mDeviceRegId = nil;
                NSLog(@"단말앱 해제 성공");
            } else {
                NSLog(@"에러 코드: %d, 사유: %@", res.rc, res.cmt);
            }
        }];
    } else {
        NSLog(@"단말앱 해제 에러: %@", resultMsg);
    }
}

#pragma mark - MessageDelegate

- (void)onDownstreamMsgRecieved:(NSArray *)msgList {
    NSLog(@"onDownstreamMsgRecieved");
}

- (void)onP2PMsgRecieved:(NSArray *)msgList {
    NSLog(@"onP2PMsgRecieved");
}

- (void)onPushMsgRecieved:(int32_t)msgSize msgList:(NSArray *)msgList alarmIdx:(int32_t)alarmIdx {
    NSLog(@"onPushMsgRecieved msg size = %d", msgSize);
    
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    [dateFormatter setDateFormat:@"yyyy/MM/dd HH:mm:ss"];
    
    for (NSDictionary *msg in msgList) {
        //double sendTime = [msg[@"serverTime"] doubleValue];
        //NSDate *date = [NSDate dateWithTimeIntervalSince1970:sendTime / 1000];
        id msgData = msg[@"data"];
        NSString *customPushStr = [NSString stringWithFormat:@"%@", msgData];
        NSLog(@"push msg = %@", customPushStr);
    }
}

- (void)onUpstreamMsgResult:(int32_t)resultCode resultMsg:(NSString *)resultMsg msgId:(NSString *)msgId {
    NSLog(@"onMsgRecieved msg upstream message id = %@", msgId);
}

- (void)onP2PMsgResult:(int32_t)resultCode resultMsg:(NSString *)resultMsg msgId:(NSString *)msgId {
    NSLog(@"onMsgRecieved msg p2p message id = %@", msgId);
}

@end

File: device-app/ios-swift/helloUms/hello/ViewController.swift
//
//  ViewController.swift
//  hello
//
//  Created by Mac on 2023/12/21.
//

import UIKit;

import RationalOwl;



class ViewController: UIViewController , DeviceRegisterResultDelegate, MessageDelegate {
    
    var mDeviceRegId: String?
    

    override func viewDidLoad() {
        super.viewDidLoad()
        
        // set device app register delegate
        let mgr: MinervaManager = MinervaManager.getInstance();
        mgr.setDeviceRegisterResultDelegate(self);
        
        // set message delegate
        mgr.setMessageDelegate(self);
    }


    @IBAction func btnReg(_ sender: Any) {
        //let gateHost: String = "gate.rationalowl.com";  // cloud
        let gateHost: String = "dev.rationalowl.com";
        let serviceId: String = "adbce9791c2c46cca10aa527a86dd6e7";
        let mgr: MinervaManager = MinervaManager.getInstance();
        mgr.registerDevice(gateHost, serviceId: serviceId, deviceRegName: "my Ios helloUms app");
    }
    
    
    @IBAction func btnUnreg(_ sender: Any) {
        let serviceId: String = "adbce9791c2c46cca10aa527a86dd6e7";
        let mgr: MinervaManager = MinervaManager.getInstance();
        mgr.unregisterDevice(serviceId);
    }
    
    
    /////////////////////////////////////////////////////////////////
    // device register delegate
    /////////////////////////////////////////////////////////////////
    
    func onRegisterResult(_ resultCode: Int32, resultMsg: String!, deviceRegId: String!) {
        print("onRegisterResult \(resultCode)")
        let msg = "\(resultMsg) registration id: \(deviceRegId)"
        print(msg)
        
        if resultCode == RESULT_OK || resultCode == RESULT_DEVICE_ALREADY_REGISTERED {
            print("rationalOwl register success!!!")
            let fDeviceRegId = deviceRegId
            
            // 변경된 API 호출 방식
            UmsApi.callInstallUmsApp(accountId: "923a0aac-abf4-493d-9f9d-787569ac53bc", deviceRegId: deviceRegId, phoneNum: nil, appUserId: nil, name: nil) { result in
                switch result {
                case .success(let data):
                    if let responseString = String(data: data, encoding: .utf8) {
                        print("응답 데이터: \(responseString)")
                    }
                    
                    do {
                        let res = try JSONDecoder().decode(PushAppProto.PushAppInstallRes.self, from: data)
                        if res.rc == RESULT_OK {
                            self.mDeviceRegId = fDeviceRegId
                            print("단말앱 등록 성공")
                        }
                        else {
                            print("에러 코드: \(res.rc), 사유: \(res.cmt)")
                        }
                    }
                    catch {
                        print("디코딩 오류: \(error.localizedDescription)")
                    }
                    
                case .failure(let error):
                    print("API 호출 실패: \(error.localizedDescription)")
                }
            }
        }
       else {
            print("단말앱 등록 에러: \(String(describing: resultMsg))")
        }
    }

    
    func onUnregisterResult(_ resultCode: Int32, resultMsg: String!) {
        if resultCode == RESULT_OK {
            guard let regId = mDeviceRegId else {
                print("등록 아이디 없음")
                return
            }
            
            // 변경된 언레지스터 호출
            UmsApi.callUnregisterUmsApp(accountId: "923a0aac-abf4-493d-9f9d-787569ac53bc", deviceRegId: regId) { result in
                switch result {
                case .success(let data):
                    do {
                        let res = try JSONDecoder().decode(PushAppProto.PushAppUnregUserRes.self, from: data)
                        if res.rc == RESULT_OK {
                            self.mDeviceRegId = nil
                            print("단말앱 해제 성공")
                        }
                        else {
                            print("에러 코드: \(res.rc), 사유: \(res.cmt)")
                        }
                    }
                    catch {
                        print("디코딩 오류: \(error.localizedDescription)")
                    }
                    
                case .failure(let error):
                    print("API 호출 실패: \(error.localizedDescription)")
                }
            }
        }
       else {
            print("단말앱 해제 에러: \(String(describing: resultMsg))")
        }
    }
    
    /////////////////////////////////////////////////////////////////
    // message delegate
    /////////////////////////////////////////////////////////////////
    
    //realtime downstream received.
    func onDownstreamMsgRecieved(_ msgList: [Any]!) {
        // hello world app don't treat realtime message
        print("onDownstreamMsgRecieved");
    }
    
    // realtime p2p received
    func onP2PMsgRecieved(_ msgList: [Any]!) {
        // hello world app don't treat realtime message
        print("onP2PMsgRecieved");
    }
    
    // while app running this callback called when push arried,
    // If there's some un-delivered push message exist, this callback delivered un-delivered push list.
    func onPushMsgRecieved(_ msgSize: Int32, msgList: [Any]!, alarmIdx: Int32) {
        print("onPushMsgRecieved msg size = \(msgSize)")
        var msg: Dictionary<String, Any>;
        // msg send time
        var sendTime: Double;
        // msgData sould be simple string or dictionary(json string on custom push)
        var msgData: Any?;
        let dateFormatter = DateFormatter();
        dateFormatter.dateFormat = "yyyy/MM/dd HH:mm:ss";
        
        for i in 0..<msgList.count {
            msg = msgList[i] as! Dictionary<String, Any>;
            // message sent time
            sendTime = msg["serverTime"] as! Double;
            let date = Date(timeIntervalSince1970: sendTime/1000);
            msgData = msg["data"];
            let customPushStr: String = (String(describing: msgData));
            print("push msg = \(customPushStr)")
        }
    }
    
    
    func onUpstreamMsgResult(_ resultCode: Int32, resultMsg: String!, msgId: String!) {
        print("onMsgRecieved msg upstream message id = " + msgId);
    }
    
    
    func onP2PMsgResult(_ resultCode: Int32, resultMsg: String!, msgId: String!) {
        print("onMsgRecieved msg p2p message id = " + msgId);
    }
}

